hpe-breckenridge/artifact-registry-yum-plugin
# Add zypper plugin

Provides the `artifact-registry` zypper plugin for resolving URLs to repositories hosted in GCP Artifact Registry.

To use this plugin, zypper repository URIs must be specified as `plugin:artifact-registry?project=PROJECT&location=LOCATION1&repository=REPOSITORY&service_account_json=PATH`.

To test on Apple M2 Max MacBook Pro (running Sonoma 14.0) against a repository containing x86_64 RPMs:

1. Cross-compile `ar-token` for linux/amd64:

    ```
    $ env GOOS=linux GOARCH=amd64 go build -o ar-token cmd/ar-token/main.go
    ```

2. Initialize docker environment via colima:

    ```
    $ colima start -a x86_64 -t vz --vz-rosetta
    ```

3. Run an opensuse/tumbleweed container with working tree mounted at /data:

    ```
    $ docker run -ti --rm -v $(pwd):/data opensuse/tumbleweed /bin/bash
    Unable to find image 'opensuse/tumbleweed:latest' locally
    latest: Pulling from opensuse/tumbleweed
    fc18dad756fe: Pull complete 
    Digest: sha256:f65e67248adef8ead6f238befc878f4680b0f82bf952e4022360623062f4f366
    Status: Downloaded newer image for opensuse/tumbleweed:latest
    85670253b6f3:/ # cd /data
    85670253b6f3:/data #
    ```

4. Install `ar-token` to `/usr/libexec`:

    ```
    85670253b6f3:/data # install -p -m 0755 ar-token /usr/libexec/
    ```

5. Verify service account key works:

    ```
    85670253b6f3:/data # /usr/libexec/ar-token --service_account_json=/data/service-account.json >/dev/null
    DEBUG: Obtain credentials using service account JSON
    DEBUG: Got a token
    ```

5. Install python311-zypp-plugin (the plugin uses `python3`):

    ```
    85670253b6f3:/data # zypper in python311-zypp-plugin
    ```

6. Install the plugin:

    ```
    85670253b6f3:/data # install -p -m 0755 zypper/artifact-registry.py /lib/zypp/plugins/urlresolver/artifact-registry
    ```

7. Add repository:

    ```
    85670253b6f3:/data # zypper ar 'plugin:artifact-registry?project=stellar-code-398917&location=us-central1&repository=devops-dev-rpm&service_account_json=/data/service-account.json' devops-dev
    ```

8. Verify by listing repository packages:

    ```
    85670253b6f3:/data # zypper pa -r devops-dev
    Loading repository data...
    Reading installed packages...
    S | Repository | Name  | Version          | Arch
    --+------------+-------+------------------+-------
      | devops-dev | gedit | 40.1-150400.2.15 | x86_64
      | devops-dev | nano  | 7.2-1.4          | x86_64
    ```
